### PR TITLE
wire kubernetes-discovery using the generic api server

### PR DESCRIPTION
**Pull request by deads2k**
This keeps compatibility with old `kubernetes-discovery` behavior by having a legacy mode, but wires up the "normal" flow of `kubernetes-discovery` to the generic API server with the presumption that we'll separately get the API resources hosted and bring in the summarization and proxying features.

I've kept the initial impl as small as I could with etcd options and secure serving options to start us off.

@kubernetes/sig-api-machinery @lavalamp @nikhiljindal @sttts

### CODE REVIEW

**Comment by deads2k** [comment]
Oh bazel how you vex me.

Let's assume I slave for the machine and make it happy.

File: cmd/kubernetes-discovery/main.go [hunk 1]
```diff
@@ -17,21 +17,23 @@ limitations under the License.
...
+	if len(os.Getenv("GOMAXPROCS")) == 0 {
+		runtime.GOMAXPROCS(runtime.NumCPU())
 	}
-	err = s.Run(port)
-	if err != nil {
-		glog.Fatalf("%v\n", err)
+
+	cmd := server.NewCommandStartDiscoveryServer(os.Stdout, os.Stderr)
+	if err := cmd.Execute(); err != nil {
+		os.Exit(1)
```
Comment (sttts) [inline cmd/kubernetes-discovery/main.go]: No printing?

File: hack/local-up-cluster.sh [hunk 2]
```diff
@@ -504,6 +510,26 @@ EOF
...
+    ${CONTROLPLANE_SUDO} "${GO_OUT}/kubernetes-discovery" \
+      --cert-dir="${CERT_DIR}" \
+      --client-ca-file="${CERT_DIR}/client-ca-bundle.crt" \
+      --bind-address="${API_BIND_ADDR}" \
+      --secure-port="${DISCOVERY_SECURE_PORT}" \
+      --tls-ca-file="${ROOT_CA_FILE}" \
+      --etcd-servers="http://${ETCD_HOST}:${ETCD_PORT}"  >"${DISCOVERY_SERVER_LOG}" 2>&1 &
+    DISCOVERY_PID=$!
+
+    # Wait for kubernetes-discovery to come up before launching the rest of the components.
```
Comment (sttts) [inline hack/local-up-cluster.sh]: why?

File: cmd/kubernetes-discovery/pkg/apiserver/apiserver.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	// apiGroupInfo.GroupMeta.GroupVersion = schema.GroupVersion{Group: "apifederation.k8s.io", Version: "v1alpha1"}
+
+	// v1alpha1storage := map[string]rest.Storage{}
+	// // v1alpha1storage["apiservices"] = apiserverstorage.NewREST(c.RESTOptionsGetter.NewFor(apifederation.Resource("apiservices")))
+
+	// apiGroupInfo.VersionedResourcesStorageMap["v1alpha1"] = v1alpha1storage
+
+	// if err := s.GenericAPIServer.InstallAPIGroup(&apiGroupInfo); err != nil {
+	// 	return nil, err
+	// }
```
Comment (sttts) [inline cmd/kubernetes-discovery/pkg/apiserver/apiserver.go]: So no API group yet. Pure bootstrapping, right?

Comment (deads2k) [inline cmd/kubernetes-discovery/pkg/apiserver/apiserver.go]: > So no API group yet. Pure bootstrapping, right?

Correct.  I'd like to merge the API pull today.  I'll comment in it.

File: hack/local-up-cluster.sh (hunk 2 above)
Comment (deads2k) [inline hack/local-up-cluster.sh]: > why?

Makes it easier to fail on it.  This is just to start it up, but I'll need to get this up in a pod later, with kube proxy pointing to its service, and then I'd like to generate the bootstrap kubeconfig to point to it.

**Comment by lavalamp** [comment]
Maybe @nikhiljindal has an opinion, but I don't think there's any production users, so I don't think you need to keep the old behavior if you're changing something.

File: cmd/kubernetes-discovery/pkg/legacy/doc.go [hunk 4]
```diff
@@ -17,4 +17,4 @@ limitations under the License.
 // Package discoverysummarizer contains code for the discovery summarizer
```
Comment (sttts) [inline cmd/kubernetes-discovery/pkg/legacy/doc.go]: s/discoverysummarizer/legacy/

**Comment by sttts** [comment]
one comment, otherwise lgtm

**Comment by deads2k** [comment]
@k8s-bot cvm gce e2e test this: issue #37940

**Comment by deads2k** [comment]
rolled into https://github.com/kubernetes/kubernetes/pull/37561 to avoid conflicts

**Pull request closed without merging**